src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,290 @@
+
+    intToBytesLE(bb, 0) // placeholder for offset
+    intToBytesLE(bb, headerLength)
+    intToBytesLE(bb, nVariants.toUInt.intRep)
+    intToBytesLE(bb, nSamples)
+    bb ++= magicNumbers
+    intToBytesLE(bb, flags)
+
+    intToBytesLE(bb, 0) // placeholder for length of sample ids
+    intToBytesLE(bb, nSamples)
+
+    var i = 0
+    var sampleBlockLength = 8
+    while (i < nSamples) {
+      sampleBlockLength += stringToBytesWShortLength(bb, sampleIds(i))
+      i += 1
+    }
+
+    val offset = headerLength + sampleBlockLength
+    updateIntToBytesLE(bb, offset, 0)
+    updateIntToBytesLE(bb, sampleBlockLength, 24)
+
+    bb.result()
+  }
+
+  def emitVariantBlock(v: Variant, va: Annotation, gs: Iterable[Genotype],
+    rsidQuery: Option[Querier], varidQuery: Option[Querier], nSamples: Int, nBitsPerProb: Int): Array[Byte] = {
+
+    val bb = new ArrayBuilder[Byte]
+    val nAlleles = v.nAlleles
+
+    require(nAlleles <= 0xffff, s"Maximum number of alleles per variant is ${ 0xffff }. Found ${ v.nAlleles }.")
+
+    val varid = varidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(v.toString)
+    val rsid = rsidQuery.flatMap(q => Option(q(va))).map(_.asInstanceOf[String]).getOrElse(".")
+
+    stringToBytesWShortLength(bb, varid)
+    stringToBytesWShortLength(bb, rsid)
+    stringToBytesWShortLength(bb, v.contig)
+    intToBytesLE(bb, v.start)
+    shortToBytesLE(bb, nAlleles)
+    stringToBytesWIntLength(bb, v.ref)
+    v.altAlleles.foreach(a => stringToBytesWIntLength(bb, a.alt))
+
+    val gtDataBlockStart = bb.length
+    intToBytesLE(bb, 0) // placeholder for length of compressed data
+    intToBytesLE(bb, 0) // placeholder for length of uncompressed data
+
+    val uncompressedData = emitGenotypeData(gs, nSamples, nAlleles, nBitsPerProb)
+
+    val uncompressedLength = uncompressedData.length
+    val compressedLength = compress(bb, uncompressedData)
+
+    updateIntToBytesLE(bb, compressedLength + 4, gtDataBlockStart)
+    updateIntToBytesLE(bb, uncompressedLength, gtDataBlockStart + 4)
+
+    bb.result()
+  }
+
+  def emitGenotypeData(gs: Iterable[Genotype], nSamples: Int, nAlleles: Int, nBitsPerProb: Int): Array[Byte] = {
+    val gtDataBuffer = new ArrayBuilder[Byte]
+    val sampleProbs = new ArrayBuilder[Int]
+
+    val nGenotypes = triangle(nAlleles)
+    val newSize = (1L << nBitsPerProb) - 1
+
+    intToBytesLE(gtDataBuffer, nSamples)
+    shortToBytesLE(gtDataBuffer, nAlleles)
+    gtDataBuffer += ploidy
+    gtDataBuffer += ploidy
+
+    val resized = new Array[Double](nGenotypes)
+    val index = new Array[Int](nGenotypes)
+
+    gs.foreach { gt =>
+      val px = gt.unboxedPX
+      if (px == null) {
+        gtDataBuffer += 0x80.toByte
+        var i = 0
+        while (i < nGenotypes - 1) {
+          sampleProbs += 0
+          i += 1
+        }
+      } else {
+        gtDataBuffer += ploidy
+        if (gt.isLinearScale)
+          resizeWeights(px, sampleProbs, resized, index, newSize)
```

Comment:
OK, I'd like to get rid of `sampleProbs`.  I would create a `BitPacker` class that has a pointer to the ArrayBuilder and has `+=(Int)` that you can write the individual probabilities.  This can be passed to `resizeWeights`  which call `BitPacker +=` rather than `sampleProbs +=`.